File: backend/app/schemas/notification.py
import uuid
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, field_validator


class NotificationResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    type: str
    title: str
    message: str
    is_read: bool
    related_entity_type: Optional[str] = None
    related_entity_id: Optional[uuid.UUID] = None
    created_at: Optional[datetime] = None

    @field_validator("id", "user_id", "related_entity_id", mode="before")
    @classmethod
    def uuid_to_str(cls, v):
        return str(v) if isinstance(v, uuid.UUID) else v

    class Config:
        from_attributes = True


class NotificationSend(BaseModel):
    user_id: uuid.UUID
    type: str = "general"
    title: str
    message: str
    related_entity_type: Optional[str] = None
    related_entity_id: Optional[uuid.UUID] = None


class NotificationPreferenceResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    email_notifications: bool
    in_app_notifications: bool
    rent_reminders: bool
    maintenance_updates: bool
    complaint_updates: bool
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    @field_validator("id", "user_id", mode="before")
    @classmethod
    def uuid_to_str(cls, v):
        return str(v) if isinstance(v, uuid.UUID) else v

    class Config:
        from_attributes = True


class NotificationPreferenceUpdate(BaseModel):
    email_notifications: Optional[bool] = None
    in_app_notifications: Optional[bool] = None
    rent_reminders: Optional[bool] = None
    maintenance_updates: Optional[bool] = None
    complaint_updates: Optional[bool] = None

File: frontend/src/types/notification.ts
export interface Notification {
  id: string;
  user_id: string;
  type: string;
  title: string;
  message: string;
  is_read: boolean;
  related_entity_type: string | null;
  related_entity_id: string | null;
  created_at: string;
}

export interface NotificationPreference {
  id: string;
  user_id: string;
  email_notifications: boolean;
  in_app_notifications: boolean;
  rent_reminders: boolean;
  maintenance_updates: boolean;
  complaint_updates: boolean;
  created_at: string;
  updated_at: string;
}

export interface NotificationPreferenceUpdate {
  email_notifications?: boolean;
  in_app_notifications?: boolean;
  rent_reminders?: boolean;
  maintenance_updates?: boolean;
  complaint_updates?: boolean;
}

export interface NotificationSend {
  user_id: string;
  type?: string;
  title: string;
  message: string;
  related_entity_type?: string | null;
}